File: ICPC/2020 ICPC Tehran/2020ICPCTehran_E.py
import sys
input = sys.stdin.readline

# read
n, m = [int(x) for x in input().split()]
grid = [input() for i in range(n)]

cnt = 0

from collections import deque
q = deque([])
d = [[float('inf')]*m for i in range(n)]
for i in range(n):
  for j in range(m):
    if grid[i][j] == "*":
      q.append((i, j))
      d[i][j] = 0
      cnt += 1
    elif grid[i][j] == "S":
      s = (i,j)
    elif grid[i][j] == "E":
      e = (i,j)

# compute distances
di = [1,-1,0,0,1,1,-1,-1]
dj = [0,0,1,-1,-1,1,1,-1]
while len(q) > 0:
  ci, cj = q.popleft()

  for k in range(8):
    x = ci+di[k]
    y = cj+dj[k]
    if x < 0 or x >= n or y < 0 or y >= m:
      continue
    if d[x][y] < float('inf'):
      continue
    d[x][y] = d[ci][cj]+1
    q.append((x,y))


vis = [[False]*m for i in range(n)]
nd = [[float('inf')]*m for i in range(n)]
si, sj = s
vis[si][sj] = True
found = False

from heapq import heappush, heappop
pq = []
heappush(pq, (-d[si][sj], si, sj))
while (len(pq) > 0):
  w, ci, cj = heappop(pq)

  if (ci, cj) == e:
    found = True

  for k in range(4):
    x = ci+di[k]
    y = cj+dj[k]
    if x < 0 or x >= n or y < 0 or y >= m:
      continue
    if vis[x][y] or grid[x][y] in "#*":
      continue
    
    vis[x][y] = True
    nd[x][y] = min(nd[ci][cj], d[x][y])
    heappush(pq, (-d[x][y], x, y))

ei, ej = e
if not found: print(-1)
elif cnt == 0: print("safe")
else: print(nd[ei][ej])

File: ICPC/2020 ICPC Tehran/2020ICPCTehran_E.cpp
#include<bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef long double ld;

int di[] = {1,-1,0,0,1,1,-1,-1};
int dj[] = {0,0,1,-1,1,-1,1,-1};
const int INF = 1e9;
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m; cin >> n >> m;
    vector<string> grid(n);
    queue<pair<int,int>> q;
    vector<vector<int>> d(n, vector<int>(m, INF));

    int cnt = 0;
    int si, sj, ti, tj;
    for (int i = 0; i < n; i++) {
        cin >> grid[i];
        for (int j = 0; j < m; j++) {
            if (grid[i][j] == '*') {
                q.push({i, j});
                d[i][j] = 0;
                cnt++;
            }
            else if (grid[i][j] == 'S') {
                si = i; sj = j;
            }
            else if (grid[i][j] == 'E') {
                ti = i; tj = j;
            }
        }
    }

    while(!q.empty()) {
        auto [ci, cj] = q.front();
        q.pop();

        for (int k = 0; k < 8; k++) {
            int x = ci+di[k];
            int y = cj+dj[k];
            if (x < 0 || x >= n || y < 0 || y >= m) continue;
            if (d[x][y] < INF) continue;

            d[x][y] = d[ci][cj]+1;
            q.push({x,y});
        }
    }    


    int ans = INF;
    vector<vector<bool>> vis(n, vector<bool>(m, 0));
    vis[si][sj] = 1;
    priority_queue<tuple<int,int,int>> pq;
    pq.push({d[si][sj], si, sj});
    bool found = false;
    while(!pq.empty()) {
        auto [w, ci, cj] = pq.top();
        pq.pop();
        ans = min(ans, d[ci][cj]);
        if (ci == ti && cj == tj) {
            found = true;
            break;
        }

        for (int k = 0; k < 4; k++) {
            int x = ci+di[k];
            int y = cj+dj[k];
            if (x < 0 || x >= n || y < 0 || y >= m) continue;
            if (vis[x][y]) continue;
            if (grid[x][y] == '#' || grid[x][y] == '*') continue;

            vis[x][y] = true;
            pq.push({d[x][y], x,y});
        }
    }

    if (!found) cout << -1 << endl;
    else if (cnt == 0) cout << "safe\n";
    else cout << ans << "\n";

    return 0;
}